File: src/main/java/org/hisp/dhis/expression/spi/ExpressionData.java
package org.hisp.dhis.expression.spi;

import lombok.Builder;
import lombok.Value;

import java.util.List;
import java.util.Map;

/**
 * The value context used during expression evaluation.
 *
 * These are the values plugged into the expression for data items, variables or the general context.
 */
@Value
@Builder(toBuilder = true)
public class ExpressionData {

    @Builder.Default
    Map<String, ? extends VariableValue> programRuleVariableValues = Map.of();
    @Builder.Default
    Map<String, Object> programVariableValues = Map.of();
    @Builder.Default
    Map<String, List<String>> supplementaryValues = Map.of();
    @Builder.Default
    Map<DataItem, Object> dataItemValues = Map.of();
    @Builder.Default
    Map<String, Object> namedValues = Map.of();
}

File: src/commonMain/kotlin/org/hisp/dhis/lib/expression/spi/ExpressionData.kt
package org.hisp.dhis.lib.expression.spi

/**
 * The value context used during expression evaluation.
 *
 *
 * These are the values plugged into the expression for data items, variables or the general context.
 */
data class ExpressionData(
    val programRuleVariableValues: Map<String, VariableValue>,
    val programVariableValues: Map<String, Any>,
    val supplementaryValues: Map<SupplementaryKey, List<String>>,
    val dataItemValues: Map<DataItem, Any>,
    val namedValues: Map<String, Any>
) {
    constructor() : this(mapOf(), mapOf(), mapOf(), mapOf(), mapOf())
}
